### PR TITLE
Mul 230/vendor page

### PR DESCRIPTION
**Pull request by rafalmal**

### CODE REVIEW

File: src/Provider/TaxonProvider.php [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+final class TaxonProvider
+{
+    private TaxonRepositoryInterface $taxonRepository;
+
+    public function __construct(TaxonRepositoryInterface $taxonRepository)
+    {
+        $this->taxonRepository = $taxonRepository;
+    }
+
+    public function provideForVendorPage(?string $slug, string $locale): ?TaxonInterface
```
Comment (senghe) [inline src/Provider/TaxonProvider.php:19]: It has no integration tests. Additionally, the query builder part has to be put into repository.

File: src/Repository/ProductRepository.php [hunk 2]
```diff
@@ -26,70 +26,27 @@ public function save(ProductInterface $product): void
...
-        Request $request,
-        ChannelInterface $channel
-    ): Pagerfanta {
-        $qb = $this->createListQueryBuilder($request->get('_locale'))
+        ChannelInterface $channel,
+        TaxonInterface $taxon,
+        string $locale,
+        array $sorting = [],
+        bool $includeAllDescendants = false,
+        ): QueryBuilder {
```
Comment (senghe) [inline src/Repository/ProductRepository.php:36]: Formattings

File: src/Repository/ProductRepository.php [hunk 3]
```diff
@@ -26,70 +26,27 @@ public function save(ProductInterface $product): void
...
+            $locale,
+            $sorting,
+            $includeAllDescendants,
+        );
+
+        $vendorId = $vendor->getId();
+
+        return $qb
             ->andWhere('o.vendor = :vendor')
-            ->setParameter('vendor', $vendor)
```
Comment (senghe) [inline src/Repository/ProductRepository.php:39]: Removing all the code here won't break any other functionality, will it?

File: src/Repository/ProductRepositoryInterface.php [hunk 4]
```diff
@@ -12,22 +12,22 @@
...
         VendorInterface $vendor,
-        Request $request,
-        ChannelInterface $channel
-    ): Pagerfanta;
+        ChannelInterface $channel,
+        TaxonInterface $taxon,
+        string $locale,
+        array $sorting = [],
+        bool $includeAllDescendants = false,
+        ): QueryBuilder;
```
Comment (senghe) [inline src/Repository/ProductRepositoryInterface.php:32]: Formatting

File: src/Resources/config/doctrine/VendorBackgroundImage.orm.xml [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    <mapped-superclass name="BitBag\OpenMarketplace\Entity\VendorBackgroundImage" table="bitbag_open_marketplace_vendor_background_image">
+        <id name="id" type="integer">
+            <generator strategy="AUTO" />
+        </id>
+
+        <field name="uuid" type="uuid" unique="true" />
+        <field name="path" column="path" type="text" />
+
+        <one-to-one field="owner" target-entity="BitBag\OpenMarketplace\Entity\VendorInterface" inversed-by="backgroundImage">
+            <join-column name="owner_id" referenced-column-name="id" nullable="false" on-delete="CASCADE" />
```
Comment (senghe) [inline src/Resources/config/doctrine/VendorBackgroundImage.orm.xml:16]: I think I would write a test for removing vendor's image, to prove it won't remove the owners entity.

File: src/Resources/config/grids/shop/sylius_shop_vendor_product.yml [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+              includeAllDescendants: "expr:parameter('sylius_shop.product_grid.include_all_descendants')"
+      sorting:
+        position: asc
+      limits: [9, 18, 27]
+      fields:
+        createdAt:
+          type: datetime
+          sortable: ~
+        position:
+          type: string
```
Comment (senghe) [inline src/Resources/config/grids/shop/sylius_shop_vendor_product.yml:25]: What's the square character?

File: src/Resources/config/grids/shop/sylius_shop_vendor_product.yml [hunk 7]
```diff
@@ -0,0 +1,40 @@
...

+          sortable: productTaxon.position
+        name:
+          type: string
```
Comment (senghe) [inline src/Resources/config/grids/shop/sylius_shop_vendor_product.yml:28]: The square

File: src/Resources/config/resources/product.yml [hunk 8]
```diff
@@ -0,0 +1,6 @@
+sylius_product:
+  resources:
+    product:
+      classes:
```
Comment (senghe) [inline src/Resources/config/resources/product.yml:4]: If we added new method to model, we should also add it's interface.

File: src/Resources/config/api_resources/VendorBackgroundImage.xml [hunk 9]
```diff
@@ -0,0 +1,72 @@
+<?xml version="1.0" ?>
+
+<!--
+This file was created by developers working at BitBag
+Do you need more information about us and what we do? Visit our https://bitbag.io website!
+We are hiring developers from all over the world. Join us and start your new, exciting adventure and become part of us: https://bitbag.io/career
+-->
+
+<resources xmlns="https://api-platform.com/schema/metadata"
```
Comment (senghe) [inline src/Resources/config/api_resources/VendorBackgroundImage.xml:9]: Do we have any tests for this?

File: src/Twig/VendorExtension.php [hunk 10]
```diff
@@ -30,16 +32,20 @@ class VendorExtension extends AbstractExtension
 
     private ChannelRepositoryInterface $channelRepository;
 
+    private TaxonRepository $taxonRepository;
```
Comment (senghe) [inline src/Twig/VendorExtension.php:35]: Interface :pray:

File: src/Twig/VendorExtension.php [hunk 11]
```diff
@@ -76,4 +83,13 @@ public function getChannel(string $channelCode): ChannelInterface
 
         return $channel;
     }
+
+    public function getMainTaxon(): TaxonInterface
```
Comment (senghe) [inline src/Twig/VendorExtension.php:87]: I would add the $channel parameter, as we could have multiple channels, and each of them can have different main taxon.

File: templates/Vendor/vendor_page.html.twig [hunk 12]
```diff
@@ -1,137 +1,73 @@
...
-                <div>
-                    {{ vendor.description}}
+                <div class="reviews-count">
+                    {% set productsAvarageRating = 0 %}
+                    {% set ratedProducts = 0 %}
+                    {% for product in vendor.products %}
+                        {% set acceptedReviews = 0 %}
+                        {% for review in product.reviews %}
+                            {% if review.status == 'accepted' %}
+                                {% set acceptedReviews = acceptedReviews + 1 %}
```
Comment (senghe) [inline templates/Vendor/vendor_page.html.twig:30]: I wouldn't calculate this here. I would put this logic into entity.

File: templates/Vendor/vendor_page.html.twig [hunk 13]
```diff
@@ -1,137 +1,73 @@
...
+                            {% endif %}
+                        {% endfor %}
+                        {% if acceptedReviews >= 1 %}
+                            {% set reviewsCount = reviewsCount + acceptedReviews %}
+                            {% set ratedProducts = ratedProducts + 1 %}
+                            {% set productsAvarageRating = productsAvarageRating + product.averageRating|round(2) %}
```
Comment (senghe) [inline templates/Vendor/vendor_page.html.twig:36]: I would put the logic into entity.

File: tests/Behat/Context/Vendor/VendorUpdateContext.php [hunk 14]
```diff
@@ -37,18 +39,22 @@ class VendorUpdateContext extends RawMinkContext
 
     private VendorImageFactoryInterface $vendorImageFactory;
 
+    private TaxonFactory $taxonFactory;
```
Comment (senghe) [inline tests/Behat/Context/Vendor/VendorUpdateContext.php:42]: Interface :pray:

File: tests/Behat/Context/Vendor/VendorUpdateContext.php [hunk 15]
```diff
@@ -211,4 +215,19 @@ public function iShouldSeeAsDefaultFormValues(
...
+    /**
+     * @Given the channel has as a menu taxon
+     */
+    public function theChannelHasAsAMenuTaxon()
+    {
+        $taxon = $this->taxonFactory->createNew();
+        $taxon->setCode('menu_category');
+        $taxon->setName('main');
+        $taxon->setSlug('main');
+        $taxon->enable();
```
Comment (senghe) [inline tests/Behat/Context/Vendor/VendorUpdateContext.php:228]: We should attach the taxon into channel. I don't see it here. Does it exist in different place? If so, why the @Given tells anything about channel?

File: tests/Integration/DataFixtures/ORM/VendorRepositoryTest/test_it_finds_correct_vendor.yml [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+    shopUser: '@user_bruce'
+    companyName: 'Wayne enterprise'
+    taxIdentifier: '1234567'
+    phoneNumber: '333111222'
+    vendorAddress: '@bruce_vendor_address'
+    slug: 'bruce-wayne-company'
+    description: 'description'
+
+
+
```
Comment (senghe) [inline tests/Integration/DataFixtures/ORM/VendorRepositoryTest/test_it_finds_correct_vendor.yml:60]:
```suggestion
```

File: tests/Integration/Repository/ProductRepositoryTest.php [hunk 17]
```diff
@@ -0,0 +1,37 @@
+<?php
```
Comment (senghe) [inline tests/Integration/Repository/ProductRepositoryTest.php:1]: No OS header

File: tests/Integration/Repository/ProductRepositoryTest.php [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+namespace Tests\BitBag\OpenMarketplace\Integration\Repository;
+
+use ApiTestCase\JsonApiTestCase;
+use BitBag\OpenMarketplace\Entity\Product;
+use BitBag\OpenMarketplace\Entity\Vendor;
+use BitBag\OpenMarketplace\Entity\VendorInterface;
+use Doctrine\ORM\QueryBuilder;
+use Sylius\Component\Core\Model\Channel;
+
+class ProductRepositoryTest extends JsonApiTestCase
```
Comment (senghe) [inline tests/Integration/Repository/ProductRepositoryTest.php:14]: Can it be final?

File: tests/Integration/Repository/VendorRepositoryTest.php [hunk 19]
```diff
@@ -0,0 +1,30 @@
+<?php
```
Comment (senghe) [inline tests/Integration/Repository/VendorRepositoryTest.php:1]: Missing OS header

File: tests/Integration/Repository/VendorRepositoryTest.php [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+class VendorRepositoryTest extends JsonApiTestCase
+{
+    public function setUp(): void
+    {
+        parent::setUp();
+        $this->entityManager = $this->getContainer()->get('doctrine.orm.entity_manager');
+        $this->repository = $this->entityManager->getRepository(Vendor::class);
+    }
+
+    public function test_it_finds_correct_vendor(): void
```
Comment (senghe) [inline tests/Integration/Repository/VendorRepositoryTest.php:20]: Too little scenarios (null scenario is missing).

File: tests/Integration/Repository/VendorRepositoryTest.php [hunk 21]
```diff
@@ -0,0 +1,30 @@
...
+
+declare(strict_types=1);
+
+namespace Tests\BitBag\OpenMarketplace\Integration\Repository;
+
+use ApiTestCase\JsonApiTestCase;
+use BitBag\OpenMarketplace\Entity\Vendor;
+use BitBag\OpenMarketplace\Entity\VendorInterface;
+
+class VendorRepositoryTest extends JsonApiTestCase
```
Comment (senghe) [inline tests/Integration/Repository/VendorRepositoryTest.php:11]: Can it be final?

**Review by senghe — CHANGES_REQUESTED** [review]

File: features/vendor/vendor_can_update_profile.feature [hunk 22]
```diff
@@ -30,7 +31,7 @@ Feature: Vendor can update his company information
     Then Pending update data should appear in database
 
   Scenario: Visiting confirmation link
-    Given There is pending update data with token value "simpletoken" for logged in vendor 
+    Given There is pending update data with token value "simpletoken" for logged in vendor
```
[the remark this answers is not in the document]
Comment (rafalmal) [reply features/vendor/vendor_can_update_profile.feature:34]: I dont understand i just removed space at the end here

File: tests/Integration/Repository/ProductRepositoryTest.php (hunk 18 above)
Comment (rafalmal) [reply tests/Integration/Repository/ProductRepositoryTest.php:21]: Added final statement in next commit.

File: tests/Integration/Repository/VendorRepositoryTest.php (hunk 21 above)
Comment (rafalmal) [reply tests/Integration/Repository/VendorRepositoryTest.php]: Added final statement in next commit.

File: src/Provider/TaxonProvider.php [hunk 23]
```diff
@@ -0,0 +1,33 @@
+<?php
+
+declare(strict_types=1);
+
+namespace BitBag\OpenMarketplace\Provider;
+
+use Sylius\Component\Taxonomy\Model\TaxonInterface;
+use Sylius\Component\Taxonomy\Repository\TaxonRepositoryInterface;
+
+final class TaxonProvider
```
[the remark this answers is not in the document]
Comment (rafalmal) [reply src/Provider/TaxonProvider.php]: Added interface.

File: src/Provider/TaxonProvider.php [hunk 24]
```diff
@@ -0,0 +1,33 @@
+<?php
```
[the remark this answers is not in the document]
Comment (rafalmal) [reply src/Provider/TaxonProvider.php:1]: Added OS header.

File: features/vendor/vendor_can_update_profile.feature [hunk 25]
```diff
@@ -10,7 +10,8 @@ Feature: Vendor can update his company information
     And the store operates in "Poland"
     And there is a "verified" vendor user "test@company.domain" registered in country "PL"
     And I am logged in as "test@company.domain"
-    
+    And the channel has as a menu taxon
```
[the remark this answers is not in the document]
Comment (rafalmal) [reply features/vendor/vendor_can_update_profile.feature:13]: FIxed typo in next commit.

File: src/Operator/VendorBackgroundOperator.php [hunk 26]
```diff
@@ -0,0 +1,52 @@
...
+use BitBag\OpenMarketplace\Factory\VendorBackgroundImageFactoryInterface;
+use Doctrine\ORM\EntityManagerInterface;
+
+final class VendorBackgroundOperator implements VendorBackgroundOperatorInterface
+{
+    private EntityManagerInterface $entityManager;
+
+    private VendorBackgroundImageFactoryInterface $vendorBackgroundImageFactory;
+
+    public function __construct(EntityManagerInterface $entityManager, VendorBackgroundImageFactoryInterface $vendorBackgroundImageFactory)
```
[the remark this answers is not in the document]
Comment (rafalmal) [reply src/Operator/VendorBackgroundOperator.php:26]: Fixed in next commit.

File: tests/Behat/Context/Vendor/VendorUpdateContext.php (hunk 14 above)
Comment (rafalmal) [reply tests/Behat/Context/Vendor/VendorUpdateContext.php]: Done in next commit.

File: src/Twig/VendorExtension.php (hunk 10 above)
Comment (rafalmal) [reply src/Twig/VendorExtension.php]: Replaced with interface.

File: src/Resources/config/grids/shop/sylius_shop_vendor_product.yml (hunk 6 above)
Comment (rafalmal) [reply src/Resources/config/grids/shop/sylius_shop_vendor_product.yml]: Removed

File: tests/Integration/Repository/VendorRepositoryTest.php (hunk 19 above)
Comment (rafalmal) [reply tests/Integration/Repository/VendorRepositoryTest.php:1]: OS header added.

File: src/Repository/ProductRepository.php (hunk 3 above)
Comment (rafalmal) [reply src/Repository/ProductRepository.php:39]: no replacing method reateListQueryBuilder with createShopListQueryBuilder covers this functionality.

File: tests/Integration/Repository/ProductRepositoryTest.php (hunk 17 above)
Comment (rafalmal) [reply tests/Integration/Repository/ProductRepositoryTest.php:1]: OS header added

File: templates/Vendor/vendor_page.html.twig (hunk 12 above)
Comment (rafalmal) [reply templates/Vendor/vendor_page.html.twig]: Moved this logic to entity.

File: templates/Vendor/vendor_page.html.twig (hunk 13 above)
Comment (rafalmal) [reply templates/Vendor/vendor_page.html.twig]: Moved this logic to entity.

File: tests/Integration/Repository/VendorRepositoryTest.php (hunk 20 above)
Comment (rafalmal) [reply tests/Integration/Repository/VendorRepositoryTest.php:27]: Added test scenario.

File: features/shop/vendor_page.feature [hunk 27]
```diff
@@ -0,0 +1,30 @@
+@vendor_page
+Feature: Displaying vendor page
```
[the remark this answers is not in the document]
Comment (rafalmal) [reply features/shop/vendor_page.feature:2]: Added scenarios.

File: src/Repository/ProductRepositoryInterface.php (hunk 4 above)
Comment (rafalmal) [reply src/Repository/ProductRepositoryInterface.php]: Fixed

File: src/Resources/config/doctrine/VendorBackgroundImage.orm.xml (hunk 5 above)
Comment (rafalmal) [reply src/Resources/config/doctrine/VendorBackgroundImage.orm.xml:16]: Added test for this case.

File: src/Resources/config/api_resources/VendorBackgroundImage.xml (hunk 9 above)
Comment (rafalmal) [reply src/Resources/config/api_resources/VendorBackgroundImage.xml:9]: Added tests.

**Pull request merged**